### PR TITLE
Improved battle obstacle selection

### PR DESCRIPTION
**Pull request by idshibanov**
Corrected obstacle to terrain mapping, more variety will be available. Updated random selection to make sure only one of a kind appear at the same time.

Tech debt: replace `std::random_shuffle` with `std::shuffle` once we move to C++11.

### CODE REVIEW

File: src/fheroes2/battle/battle_board.cpp [hunk 1]
```diff
@@ -634,128 +655,21 @@ void Battle::Board::SetCobjObjects( const Maps::Tiles & tile )
...
-            dst = GetObstaclePosition();
-        if ( objs.size() > 1 && 7 < Rand::Get( 1, 10 ) )
-            SetCobjObject( *Rand::Get( objs ), dst );
+    const size_t objectsToPlace = std::min( objs.size(), static_cast<size_t>( Rand::Get( 0, 4 ) ) );
+    std::random_shuffle( objs.begin(), objs.end() );
+
+    for ( size_t i = 0; i < objectsToPlace; ++i ) {
+        const s32 dest = GetObstaclePosition();
+        while ( at( dest ).GetObject() )
+            dest = GetObstaclePosition();
```
Comment (idshibanov) [inline src/fheroes2/battle/battle_board.cpp:664]: Might have to revert that? We're updating `dest` while searching a free spot to place the obstacle...

Comment (ihhub) [reply src/fheroes2/battle/battle_board.cpp:664]: ah, yes. You're right. I'll revert this back.

**Comment by ihhub** [comment]
Hi @Branikolog , could you please check if in the original game we could have 2 or more same objects in battlefield?

**Comment by Branikolog** [comment]
Hello, @ihhub. I've tried 50+ battles on a grass. None of them had 2 same objects at once.

**Comment by idshibanov** [comment]
> Hello, @ihhub. I've tried 50+ battles on a grass. None of them had 2 same objects at once.

Yes, I based it on my findings in the original code. It explicitly sets a boolean when an obstacle is placed. If it rolls the same one it just skips and tries again. Loop can roll up to 500 tries.

**Comment by ihhub** [comment]
@idshibanov thank you very much for your non-stop contribution!

**Pull request merged**